### PR TITLE
[ARM] Migrate resource to track2 SDK

### PR DESCRIPTION
**Pull request by zhoxing-ms**
**Description**

**This PR upgrades the Python SDK version of `azure-mgmt-resource` from `12.1.0` to `16.1.0`**

**Here are the main changes for resource migration to track2:**
1. LRO operations are renamed to `begin_xxx_xxx`. For example in track1, the name is `create_or_update`; while in track2, the name is `begin_create_or_update`.
2. When `api-version` is less than `2019-10-01`, the parameters of all deployment operations are changed from `DeploymentProperties` to `Deployment`, which is consistent with the SDK after version `2019-10-01`.
3. The objects of input parameters for some methods are further encapsulated:
- `begin_what_if()`: `DeploymentWhatIfProperties` --> `DeploymentWhatIf`
- `begin_what_if_at_subscription_scope()/...`: `DeploymentWhatIfProperties` --> `ScopedDeploymentWhatIf`
- `begin_export_template()`: flattened parameters --> `ExportTemplateRequest`
- `begin_move_resources()`: flattened parameters --> `ResourcesMoveInfo`
- `links_client.create_or_update()`: `ResourceLinkProperties` --> `ResourceLink`
- `tags.create_or_update_at_scope()`: flattened parameters --> `TagsResource`
- `tags.update_at_scope()`: flattened parameters --> `TagsPatchResource`
4. In method `get()` of `ResourcesOperations`, some parameters are passed in through `**kwargs` instead
5. `CloudError` and `HttpOperationError` are replaced by `HttpResponseError`
6. In track 2, the logic corresponding to the class name of Codegen generating operations changes, so `DeploymentOperations` becomes `DeploymentOperationsOperations`
For more details, please refer to this issue: https://github.com/Azure/autorest.python/issues/910#issuecomment-809516272
7. `config` in management client becomes a private variable `_config`.
8. In method `policy_definitions.create_or_update_at_management_group()`, the positions of parameters `management_group`and `parameters` are reversed.
9. `generate_client_request_id` in `ResourceManagementClientConfiguration` no longer needs, SDK automatically add `x-ms-request-id` now through the `RequestIdPolicy`
10. The `accept_language` is no longer supported in `ResourceManagementClientConfiguration`, so the fixed value `en-US` in track 2 is hard coded.
11. The pipeline and policy in management client migrated from `msrest` to `azure-core`. And when doing ARM deployment, the pipeline needs to add custom policy `JsonCTemplatePolicy()` to support JSON templates.
12. `get_by_id()` and `get()` of `ResourcesOperations` no longer support parameter `raw`, but passes in a callback method `cls` instead. So the custom callback function `add_response_body()` is used to realize the logic of `include_response_body`.
13. The resource action lacks generic `POST` operation in Swagger, so for the time being, `AzureOperationPoller` of `msrest`is still used for low-cost migration. And the request way is changed to `pipeline.run` in track 2.
14. Before version `2019-08-01`, `get_by_id()` and `get()` of `ResourcesOperations` lacked parameter `api-version`. The problem was solved by modifying the Swagger to regenerate a new Python SDK. PR link: [Swagger PR](https://github.com/Azure/azure-rest-api-specs/pull/13834)
15. Fix some live tests due to the changes in track2.

**Known commands that are not covered by test**
1. `appservice domain create`
2. `appservice ase create`

**Migration for extension**
PR link: https://github.com/Azure/azure-cli-extensions/pull/3355

**Live Test**
[result link](https://clitestresultstac.blob.core.windows.net/20210510-203122-491033/index.html)

**TODO**
- [ ] Refactor the temporary code in the subsequent PR https://github.com/Azure/azure-cli/issues/18023

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
ARM

File: src/azure-cli/azure/cli/command_modules/resource/custom.py
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:362]: Consider using `per_call_policies` introduced by https://github.com/Azure/azure-sdk-for-python/pull/17340.

File: src/azure-cli/azure/cli/command_modules/resource/custom.py
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:3192]: `_prepare_client_kwargs_track2` from core already configures `x-ms-client-request-id`.

https://github.com/Azure/azure-cli/blob/039fa21db4fdf6dfb22958aaf1c3f614e1fc3b59/src/azure-cli-core/azure/cli/core/commands/client_factory.py#L156-L158

**Comment by jiasli** [comment]
> Migrate for extension

I would suggest we provide a method for users to revert back to old versions of Azure CLI because these are not small changes and may break extensions, especially those not in [azure-cli-extensions](https://github.com/Azure/azure-cli-extensions) (https://github.com/Azure/azure-cli/issues/13653).

**Comment by jiasli** [comment]
Please kindly put this PR on hold until #17671 and #17812 are merged so that they can be tested together.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:1047]: For old api version, your new SDK package still use `validate` as function name? Typically with new track2 SDK, all api versions for long running operation should use begin _xxx().

File: src/azure-cli/azure/cli/command_modules/network/custom.py
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:249]: Can we get the Deployment model from SDK which api-version less than 2019-10-01

File: src/azure-cli/azure/cli/command_modules/network/custom.py
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:3338]: Same concern as above.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py:1047]: Yes, this is because before version `2019-10-01`, `DeploymentsOperations.validate` is a normal synchronous operation. And only after or equal this version, this function is the long running operation.

Swagger definition before version `2019-10-01`: [Swagger link](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/resources/resource-manager/Microsoft.Resources/stable/2019-08-01/resources.json#L259)
Swagger definition after (or equal) version `2019-10-01`: [Swagger link](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/resources/resource-manager/Microsoft.Resources/stable/2020-10-01/resources.json#L260)

File: src/azure-cli/azure/cli/command_modules/network/custom.py
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:249]: Yes, because in track 2, the SDK before version a has added this model
Related code: [code link](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/resources/azure-mgmt-resource/azure/mgmt/resource/resources/v2019_08_01/models/_models.py#L170)

**Review by fengzhou-msft — APPROVED** [review]

**Review by kairu-ms — APPROVED** [review]

**Review by jiasli — APPROVED** [review]

**Review by houk-ms — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

**Review by qwordy — APPROVED** [review]
Approve vm part

**Review by Juliehzl — APPROVED** [review]

**Pull request merged**
